src/nupic/research/connections.py

```diff
@@ -251,50 +484,93 @@ def computeActivity(self, activeInput, activePermanenceThreshold,
+                                               update the lastUsedIteration on
+                                               active segments and the internal
+                                               iteration variable
 
     @return (tuple) Contains:
                       `activeSegments`         (list),
-                      `matchingSegments`        (list),
+                      `matchingSegments`       (list),
 
     Notes:
-      activeSegments and matchingSegments are sorted by the cell they are on.
+      activeSegments and matchingSegments are sorted by the cell they are on
+      and they are lists of SegmentOverlap objects.
     """
 
-    numActiveSynapsesForSegment = [0] * self._nextSegmentIdx
-    numMatchingSynapsesForSegment = [0] * self._nextSegmentIdx
+    numActiveSynapsesForSegment = [0] * self._nextFlatIdx
+    numMatchingSynapsesForSegment = [0] * self._nextFlatIdx
 
     for cell in activeInput:
-      for synapseData in self.synapsesForPresynapticCell(cell).values():
-        segment = synapseData.segment
+      for synapse in self.synapsesForPresynapticCell(cell):
+        synapseData = self.dataForSynapse(synapse)
+        segment = synapse.segment
         permanence = synapseData.permanence
-        if permanence >= matchingPermananceThreshold:
-          numMatchingSynapsesForSegment[segment] += 1
-          if synapseData.permanence >= activePermanenceThreshold:
-            numActiveSynapsesForSegment[segment] += 1
+        segmentData = self.dataForSegment(segment)
+        if permanence - matchingPermananceThreshold > -EPSILON:
+          numMatchingSynapsesForSegment[segmentData.flatIdx] += 1
+          if permanence - activePermanenceThreshold > -EPSILON:
+            numActiveSynapsesForSegment[segmentData.flatIdx] += 1
+
+    if recordIteration:
+      self._iteration += 1
 
     activeSegments = []
     matchingSegments = []
-    for i in xrange(self._nextSegmentIdx):
-      if numActiveSynapsesForSegment[i] >= activeSynapseThreshold:
-        activeSegments.append(i)
+    for i in xrange(self._nextFlatIdx):
+      segment = self._segmentForFlatIdx[i]
+      numActive = numActiveSynapsesForSegment[i]
+      if numActive >= activeSynapseThreshold:
+        segmentOverlap = SegmentOverlap(segment, numActive)
+        activeSegments.append(segmentOverlap)
+        
+        if recordIteration:
+          self.dataForSegment(segment).lastUsedIteration = self._iteration
+
+
+    for i in xrange(self._nextFlatIdx):
+      segment = self._segmentForFlatIdx[i]
+      numMatching = numMatchingSynapsesForSegment[i]
+      if numMatching >= matchingSynapseThreshold:
+        segmentOverlap = SegmentOverlap(segment, numMatching)
+        matchingSegments.append(segmentOverlap)
+
+    segCmp = lambda a, b: (a.segment.cell - b.segment.cell
+                           if a.segment.cell - b.segment.cell != 0
+                           else a.segment.idx - b.segment.idx)
```

Comment:
This is slower than necessary since it always computes the difference twice. I don't know if it can be written fast as a lambda. 

Maybe do `sorted(activeSegments, key = lambda x: x.segment)` and then implement the `__lt__`, `__le__`, etc. methods on the `Segment` class. These methods could be written more efficiently than a lambda, since you can store the `segment.cell - segment.cell` value rather than compute it twice.